### PR TITLE
snapshotter: use syncfs system call

### PR DESCRIPTION
**Pull request by zhouhaibing089**
`sync` system call triggers a full page cache sync which may not always work, especially in kubernetes environment where it is very easy to be interfered by others. I have seen several cases where a broken nfs mount is blocking kaniko from doing its job.

With `syncfs`, it only writes cache back to disk for the current filesystem that is used by kaniko which is supposed to be more reliable.

Here is the details that I captured recently. From kaniko executor logs:

```console
/workspace # executor --no-push
INFO[0000] Retrieving image manifest ubuntu:22.04
INFO[0000] Retrieving image ubuntu:22.04 from registry index.docker.io
INFO[0001] Built cross stage deps: map[]
INFO[0001] Retrieving image manifest ubuntu:22.04
INFO[0001] Returning cached image manifest
INFO[0001] Executing 0 build triggers
INFO[0001] Building stage 'ubuntu:22.04' [idx: '0', base-idx: '-1']
INFO[0001] Unpacking rootfs as cmd RUN apt-get update requires it.
INFO[0003] RUN apt-get update
INFO[0003] Initializing snapshotter ...
INFO[0003] Taking snapshot of full filesystem...
```

It hangs there forever in `Taking snapshot of full filesystem` - see https://github.com/GoogleContainerTools/kaniko/issues/1333 as well.

I looked the stack of kaniko process and see this:

```console
root@<my-host-name>:/proc/180716/task# cat 182585/stack
[<0>] sync_inodes_sb+0x10b/0x2b0
[<0>] sync_inodes_one_sb+0x15/0x20
[<0>] iterate_supers+0xa0/0x100
[<0>] ksys_sync+0x42/0xb0
[<0>] __do_sys_sync+0xe/0x20
[<0>] do_syscall_64+0x59/0xc0
[<0>] entry_SYSCALL_64_after_hwframe+0x44/0xae
```

where `180716` is the pid of kaniko executor, and one of its subtask is waiting in `ksys_sync` forever.

At that point of time, running `sync` (command) on the host was also stuck forever. We further checked and then identified the system call was waiting for one of nfs mount to flush its superblock (but due to connectivity issues, it can't complete).

```console
root@<my-host-name>:/# dmesg -T | grep nfs
[Thu Oct 26 04:05:18 2023] FS-Cache: Netfs 'nfs' registered for caching
[Thu Oct 26 04:12:00 2023]  ? nfs_ctx_key_to_expire+0x69/0xf0 [nfs]
[Thu Oct 26 04:12:00 2023]  nfs_wb_all+0x28/0xf0 [nfs]
[Thu Oct 26 04:12:00 2023]  nfs4_file_flush+0x73/0xb0 [nfsv4]
[Thu Oct 26 04:12:28 2023] nfs: server <nfs-server-host-name> not responding, still trying
[Thu Oct 26 04:12:43 2023] nfs: server <nfs-server-host-name> not responding, still trying
[Thu Oct 26 04:14:00 2023]  ? nfs_ctx_key_to_expire+0x69/0xf0 [nfs]
[Thu Oct 26 04:14:00 2023]  nfs_wb_all+0x28/0xf0 [nfs]
[Thu Oct 26 04:14:00 2023]  nfs4_file_flush+0x73/0xb0 [nfsv4]
```

While this is technically not related to kaniko, but by switching to `syncfs` with fd from current directory, it is going to only sync the current file system which holds the provided fd, and this is not affected by external filesystems.

**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [ ] Adds integration tests if needed.

_See [the contribution guide](../CONTRIBUTING.md) for more details._

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit tests and or integration tests added.

**Release Notes**

Describe any changes here so maintainer can include it in the release notes, or delete this block.

```
snapshotter: use syncfs system call rather than sync
```

### CODE REVIEW

File: pkg/snapshot/snapshot.go [hunk 1]
```diff
@@ -155,7 +155,16 @@ func (s *Snapshotter) scanFullFilesystem() ([]string, []string, error) {
...
 	// which can lag if sync is not called. Unfortunately there can still be lag if too much data needs
 	// to be flushed or the disk does its own caching/buffering.
-	syscall.Sync()
+	dir, err := os.Open(s.directory)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer dir.Close()
+	// 306 is the number for sys_syncfs
+	_, _, errno := syscall.Syscall(306, dir.Fd(), 0, 0)
```
Comment (aaron-prindle) [inline pkg/snapshot/snapshot.go:164]: nit: Can you add some OS related checks to ensure that this is only done for operating systems that support sys_syncfs.  Perhaps look at `runtime.GOOS` and only do the sys_syncfs logic for linux OS?  Admittedly kaniko is mostly run in a container on top of linux containerization but just for completeness/future-proofing this check might make sense

**Comment by aaron-prindle** [comment]
Thanks for the PR here @zhouhaibing089, appreciate the fix!  Left one nit comment, once that is addressed we can get this merged

File: pkg/snapshot/snapshot.go (hunk 1 above)
Comment (zhouhaibing089) [reply pkg/snapshot/snapshot.go]: Sounds good. Thanks for the suggestion.

Comment (zhouhaibing089) [reply pkg/snapshot/snapshot.go]: Ah, `306` is also amd64 specific. I think I can leverage golang.org/x/sys/unix.

**Review by aaron-prindle — APPROVED** [review]
LGTM!  Thanks @zhouhaibing089!

**Comment by rayunstop** [comment]
great job, help a lot. Thanks.

**Pull request merged**